woocommerce/woocommerce
# Fix Regression of Orders API meta_data in #27492

## All Submissions:






## Changes proposed in this Pull Request:



### Problem

This fixes the failing unit test in PR #27492. The failing unit test shows that there was a regression on how I implemented the new `display_key` and `display_value` attributes (originally in https://github.com/woocommerce/woocommerce-rest-api/pull/229).

```
1) WC_Tests_API_Orders::test_create_order
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
-        0 => WC_Meta_Data Object (...)
-        1 => WC_Meta_Data Object (...)
-        2 => WC_Meta_Data Object (...)
+        0 => Array (...)
+        1 => Array (...)
```

As shown above, the expected number of metadata is 3 but the new implementation only returned 2.

The reason for this was that I was originally using the result of `$item->get_formatted_meta_data()` to build the final `$data['meta_data']` returned by the API. 

https://github.com/woocommerce/woocommerce/blob/ecb2936f9e0ffb4933aa64a108b216fbcaa78302/includes/rest-api/Controllers/Version2/class-wc-rest-orders-v2-controller.php#L201-L206

I have unfortunately missed the fact that [`get_formatted_meta_data`](https://github.com/woocommerce/woocommerce/blob/ecb2936f9e0ffb4933aa64a108b216fbcaa78302/includes/class-wc-order-item.php#L257) **can exclude** some metadata depending on their attribute values: 

https://github.com/woocommerce/woocommerce/blob/ecb2936f9e0ffb4933aa64a108b216fbcaa78302/includes/class-wc-order-item.php#L265-L267

### Solution

To restore the previous behavior as much as possible, I changed the logic for building the _final_ `$data['meta_data']` array by:

1. Primarily using the original `$data['meta_data']` array instead of `$formatted_meta_data`.
2. Merging `display_value` and `display_key` attributes to the `$data['meta_data']` if possible. 

This solution (ac4cba3) keeps the original number of metadata in the response while also applying the new `display_key` and `display_value` attribute values. 

#### Fix for Failing Test

As for the fix for the failing test, I changed the assertions (5651bef and d6600d8) so that this `$shipping->get_meta_data()` is no longer used for comparison: 

https://github.com/woocommerce/woocommerce/blob/ecb2936f9e0ffb4933aa64a108b216fbcaa78302/tests/legacy/unit-tests/rest-api/Tests/Version2/orders.php#L272-L274

This can no longer be simply compared like this because there are new `display_key` and `display_value` properties in the API response. 

## How to test the changes in this Pull Request:

The changes in #27492 (https://github.com/woocommerce/woocommerce-rest-api/pull/229) are covered by unit tests so they should still be passing even if this PR changed the logic for the `$data['meta_data']`.  

1. Please confirm that the `WC_Tests_API_Orders_V2.test_create_order()` assertions still test what it originally tested before this PR. 
2. Please confirm that the `test_create_order()` unit test passes now. This confirms that the regression was fixed.

## Other information:




## Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.
